src/Shared/BinarySerializer.cs

```diff
@@ -3,44 +3,76 @@
 // Licensed under the MIT License. See License.txt in the project root for license information.
 //
 
-using System.IO;
-using System.Runtime.Serialization.Formatters.Binary;
+using System;
+using System.Linq;
+using System.Text;
+using System.Web.SessionState;
 
 namespace Microsoft.Web.Redis
 {
+
     public class BinarySerializer : ISerializer
     {
+
+        public enum DataTypes : byte
+        {
+            redisNull,
+            initializeItem,
+            str,
+            byteArray
+        }
+
         public byte[] Serialize(object data)
         {
-            if (data == null)
-            {
-                data = new RedisNull();
-            }
-            var binaryFormatter = new BinaryFormatter();
-            using (var memoryStream = new MemoryStream())
+            switch (data)
             {
-                binaryFormatter.Serialize(memoryStream, data);
-                byte[] objectDataAsStream = memoryStream.ToArray();
-                return objectDataAsStream;
+                case null:
+                    return new byte[] { (byte)DataTypes.redisNull };
+                case SessionStateActions.InitializeItem:
+                    return new byte[] { (byte)DataTypes.initializeItem };
+                case string:
+                    return PrependHeaderToData(Encoding.UTF8.GetBytes((string)data), (byte)DataTypes.str);
+                case byte[]:
```

Comment:
Are we really seeing byte arrays passed in here? What's the scenario where that happens?

Reply:
https://github.com/Azure/aspnet-redis-providers/blob/main/test/RedisSessionStateProviderUnitTest/RedisUtilityTests.cs#L130

Reply:
Another concern: most of the cases in this switch statement are types, while `null` is a value. `switch (data)` will only switch on the value, so the type cases will never be hit under normal circumstances. Rather than a `switch`, you'll probably need `if` statements to check for null, and checks like `if (data is string)` for the types.

Reply:
Null is handled on line 70 and line 32: https://github.com/Azure/aspnet-redis-providers/blob/459da31676c2b754979c3a627ecdef1950dd7689/src/Shared/BinarySerializer.cs#L70